pawnpy/__init__.py

```diff
@@ -167,82 +181,107 @@ class AMX():
     def __init__(self, filename):
         self._filename = filename
         self._amx = AMXNative()
-        self._public_functions = dict()  # name:str -> index:number
 
         result = lib.aux_LoadProgram(
             byref(self._amx), c_char_p(filename.encode('utf-8')), 0)
         if result != AMX_ERR_NONE:
-            raise Exception('aux_LoadProgram failed with code %d' % result)
-
-        numNatives = c_int()
-        result = lib.amx_NumNatives(byref(self._amx), byref(numNatives))
-        if result != AMX_ERR_NONE:
-            raise Exception('amx_NumNatives failed with code %d' % result)
+            raise RuntimeError('aux_LoadProgram failed with code %d' % result)
 
-        print(numNatives)
+        self._publics = None
+        self.__init_publics()
 
-        if numNatives.value != 0:
-            name = create_string_buffer(sNAMEMAX + 1)
-            result = lib.amx_GetNative(byref(self._amx), 0, name, None)
-            if result != AMX_ERR_NONE:
-                raise Exception('amx_GetNative failed with code %d' % result)
+        self._natives = None
+        self.__init_natives()
 
-            print(name.value)
+        self._callbackptr = AMX_CALLBACK(self._callback)
+        lib.amx_SetCallback(self._amx, self._callbackptr)
 
-        self.__read_all_public_functions()
+    def __init_publics(self):
+        """Fills publics info and hooks them up to AMX attrs"""
 
-    def __read_all_public_functions(self):
-        """Reads all public functions and store name -> index in self._public_functions"""
+        setattr(self, 'main', lambda *args: self._exec(-1, *args))
 
-        num_pub_func = c_int()
-        err_code = lib.amx_NumPublics(byref(self._amx), byref(num_pub_func))
+        num_publics = c_int()
+        err_code = lib.amx_NumPublics(byref(self._amx), byref(num_publics))
         if AMX_ERR_NONE != err_code:
-            raise RuntimeError(
-                "Error calling getting public functions count, code %d" % err_code)
+            raise RuntimeError('amx_NumPublics failed with code %d' % result)
+
+        self._publics = []
 
-        for i in range(0, num_pub_func.value):
+        for i in range(num_publics.value):
             name = create_string_buffer(sNAMEMAX + 1)
 
             err_code = lib.amx_GetPublic(byref(self._amx), i, name, None)
             if AMX_ERR_NONE != err_code:
                 raise RuntimeError(
-                    "Can't read name of %d public function, code %d" % (i, err_code))
+                    'amx_NumPublics failed with code %d' % result)
+
+            name = codecs.decode(name.value)
+            self._publics.append(name)
+
+            print('Public function %s' % name)
```

Comment:
Do we really need to print this? May be has some util function to print info about AMX?

Reply:
Removed - it's debug